core/src/services/zookeeper/backend.rs

```diff
@@ -0,0 +1,305 @@
+                let auth = [format!("{user}:").as_bytes().to_vec(), password].concat();
+                (auth, zk::Acl::creator_all())
+            }
+            _ => {
+                // TODO: change to warn
+                warn!("username and password isn't set, default use `anyone` acl");
+                (Vec::<u8>::new(), zk::Acl::anyone_all())
+            }
+        };
+
+        Ok(ZookeeperBackend::new(ZkAdapter {
+            endpoint,
+            auth,
+            acl,
+            client: OnceCell::new(),
+        }))
+    }
+}
+
+/// Backend for Zookeeper service
+pub type ZookeeperBackend = kv::Backend<ZkAdapter>;
+
+#[derive(Clone)]
+pub struct ZkAdapter {
+    endpoint: String,
+    auth: Vec<u8>,
+    client: OnceCell<zk::Client>,
+    acl: &'static [zk::Acl],
+}
+
+impl Debug for ZkAdapter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Adapter");
+        ds.field("endpoint", &self.endpoint);
+        ds.field("acl", &self.acl);
+        ds.finish()
+    }
+}
+
+impl ZkAdapter {
+    async fn get_connection(&self) -> Result<zk::Client> {
+        if let Some(client) = self.client.get() {
+            return Ok(client.clone());
+        }
+        match zk::Client::connect(&self.endpoint.clone()).await {
+            Ok(client) => {
+                if !self.auth.is_empty() {
+                    client
+                        .auth(ZOOKEEPER_AUTH_SCHEME.to_string(), self.auth.clone())
+                        .await
+                        .map_err(parse_zookeeper_error)?;
+                }
+                self.client.set(client.clone()).ok();
+                Ok(client)
+            }
+            Err(e) => Err(Error::new(ErrorKind::Unexpected, "error from zookeeper").set_source(e)),
+        }
+    }
+
+    fn create_nested_node<'a>(
+        &'a self,
+        path: &'a str,
+        value: &'a [u8],
+    ) -> BoxFuture<'a, Result<()>> {
+        eprintln!("{path}");
+        let mut path = path.to_string();
+        async move {
+            return match self
+                .get_connection()
+                .await?
+                .create(
+                    &path,
+                    value,
+                    &zk::CreateOptions::new(zk::CreateMode::Persistent, self.acl),
+                )
+                .await
+            {
+                Ok(_) => Ok(()),
+                Err(e) => match e {
+                    zk::Error::NoNode => {
+                        let idx = path.rfind('/').unwrap();
+                        let tmpath = path.clone();
+                        path = path.to_string().substring(0, idx).to_string();
+                        if path.as_bytes()[0] != b'/' {
+                            path = "/".to_string() + path.strip_suffix('/').unwrap_or(&path);
+                        }
+                        match self.create_nested_node(&path, value).await {
```

Comment:
If it's required to create parent node first for zookeepr, I perfer to check if parent exist first instead handling error.

Reply:
Yes, `create_nested_node()` is a recursive function that creates parent node recursively till a node exists. So when we find that the parent node doesn't exit (the NoNode error in zookeeper-client indicates this) we call this function to create the parent nodes till all the parent nodes on the path in the tree has created. Also it's because the fact that the function is both recursive and async that we use `BoxFuture`.